value/maybe.go

```diff
@@ -0,0 +1,80 @@
+package value
+
+import "fmt"
+
+// Maybe is a container that can hold a value of type T, which may either be
+// present or absent. A zero value is ready for use, and is absent.
+//
+// It is safe to copy and assign a Maybe value, but note that if a value is
+// present, only a shallow copy of the underlying value is made. Maybe values
+// are comparable if and only if T is comparable.
+type Maybe[T any] struct {
+	value   T
+	present bool
+}
+
+// Just returns a Maybe whose present value is v.
+func Just[T any](v T) Maybe[T] { return Maybe[T]{value: v, present: true} }
+
+// Check returns a present Maybe with value v if err == nil; otherwise it
+// returns absent.
+func Check[T any](v T, err error) Maybe[T] {
+	if err == nil {
+		return Just(v)
+	}
+	return Maybe[T]{}
+}
+
+// Absent returns an absent Maybe of the specified type.  It is a legibility
+// notation equivalent to the zero value of type Maybe[T].
+func Absent[T any]() Maybe[T] { return Maybe[T]{} }
+
+// Present reports whether a value is present in m.
+func (m Maybe[T]) Present() bool { return m.present }
+
+// GetOK reports whether a value is present in m, and if so returns that value.
+// If a value is not present, GetOK returns the zero of T.
+func (m Maybe[T]) GetOK() (T, bool) { return m.value, m.present }
+
+// Get returns the value present in m, if any; or else the zero value.
+func (m Maybe[T]) Get() T { return m.value }
+
+// Or returns the value of m if a value is present, otherwise it returns o.
+func (m Maybe[T]) Or(o T) T {
```

Comment:
It's funny you mention having it return `Maybe[T]`. That is how I originally wrote it, but then I talked myself out of it on the grounds that you're basically always going to unwrap it anyway.

But your point about the value on the LHS vs. the RHS of the function makes sense, and being able to write, e.g.,

```go
return someThing.Or(24)
```

would not be possible with the pure value construction. I can see the appeal of being explicit here, and the short-circuit framing is nice.

I'm not extremely attached to the name `Or`. I also considered `OrElse` but I'm not sure that really mitigates the problem either.

Reply:
Followup thought: I recalled that there was a proposal to introduce a `zero` builtin identifier to represent an untyped zero value. That would enable `foo.Get(zero)` to merge the "return zero value" and "return a default of my choice" APIs.

Alas, zero is not happening for the time being :( https://github.com/golang/go/issues/61372#issuecomment-1731525247

Reply:
I did consider making a `value.Zero()` but Go's type inference is just a little too strict for that to be worth it; you wind up having to write the type constraint explicitly most of the time because it doesn't do proper bidirectional unification.